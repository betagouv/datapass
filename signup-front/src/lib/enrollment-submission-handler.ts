import { getErrorMessages } from '.';
import {
  ActionConfiguration,
  EnrollmentAction,
} from './enrollment-actions-configuration';
import {
  createOrUpdateEnrollment,
  deleteEnrollment,
  computeNextEnrollmentState,
} from '../services/enrollments';

export const handleSubmissionAction = async (
  action: EnrollmentAction,
  actionConfiguration: ActionConfiguration,
  enrollment: any,
  updateEnrollment: Function,
  message?: string
): Promise<{
  redirectToHome: boolean;
  successMessages: string[];
  errorMessages: string[];
}> => {
  const resultMessages = {
    errorMessages: [] as string[],
    successMessages: [] as string[],
    redirectToHome: false,
  };

  try {
    let comment = null;

    if (actionConfiguration.promptForComment) {
      try {
        comment = message;
      } catch (e) {
        return resultMessages;
      }
    }

    let enrollmentId = enrollment.id;

    if (enrollment.acl.update) {
      const newEnrollment = await createOrUpdateEnrollment({
        enrollment,
      });
      updateEnrollment(newEnrollment);
      enrollmentId = newEnrollment.id;

      resultMessages.successMessages.push('Votre demande a été sauvegardée.');
    }

    if (action === 'update') {
      return resultMessages;
    }

    if (action === 'destroy') {
      await deleteEnrollment({ id: enrollmentId });
    }

    if (actionConfiguration.needsToComputeNextEnrollmentState) {
      await computeNextEnrollmentState({
        action,
        id: enrollmentId,
        comment,
      });
    }

    resultMessages.redirectToHome = true;

    return resultMessages;
  } catch (error) {
    resultMessages.errorMessages.push(...getErrorMessages(error));

    return resultMessages;
  }
};
