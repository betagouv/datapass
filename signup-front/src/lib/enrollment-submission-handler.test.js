import {
  EnrollmentAction,
  userInteractionsConfiguration,
} from './enrollment-actions-configuration';
jest.mock('../services/enrollments');
// eslint-disable-next-line import/first
import { handleSubmissionAction } from './enrollment-submission-handler';
// eslint-disable-next-line import/first
import {
  computeNextEnrollmentState,
  deleteEnrollment,
  createOrUpdateEnrollment,
} from '../services/enrollments';

describe('When submitting the enrollment form', () => {
  const enrollment = { id: Symbol(), acl: {} };
  const updateEnrollment = jest.fn();

  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe('with the notify action', () => {
    const action = EnrollmentAction.notify;
    const actionConfiguration = userInteractionsConfiguration.notify;

    it('calls for the enrollment state update', async () => {
      const userMessage = 'La barbe de la femme à Georges Moustaki';

      const output = await handleSubmissionAction(
        action,
        actionConfiguration,
        enrollment,
        updateEnrollment,
        userMessage,
        false
      );

      expect(computeNextEnrollmentState).toHaveBeenCalledWith({
        action: 'notify',
        comment: userMessage,
        id: enrollment.id,
      });
      expect(output).toMatchSnapshot();
    });
  });

  describe('with the destroy action', () => {
    const action = EnrollmentAction.destroy;
    const actionConfiguration = userInteractionsConfiguration.destroy;

    it('calls the delete endpoint', async () => {
      const output = await handleSubmissionAction(
        action,
        actionConfiguration,
        enrollment,
        updateEnrollment
      );

      expect(deleteEnrollment).toHaveBeenCalledWith({
        id: enrollment.id,
      });
      expect(output).toMatchSnapshot();
    });
  });

  describe('with the update action', () => {
    const action = EnrollmentAction.update;
    const actionConfiguration = userInteractionsConfiguration.update;

    const enrollmentToUpdate = { ...enrollment, acl: { update: true } };

    it('calls the update endpoint', async () => {
      createOrUpdateEnrollment.mockResolvedValue(enrollmentToUpdate);

      const output = await handleSubmissionAction(
        action,
        actionConfiguration,
        enrollmentToUpdate,
        updateEnrollment
      );

      expect(createOrUpdateEnrollment).toHaveBeenCalledWith({
        enrollment: enrollmentToUpdate,
      });
      expect(output).toMatchSnapshot();
    });

    it('displays an error if update fails', async () => {
      createOrUpdateEnrollment.mockRejectedValue("Pas d'update désolé");

      const output = await handleSubmissionAction(
        action,
        actionConfiguration,
        enrollmentToUpdate,
        updateEnrollment
      );

      expect(output).toMatchSnapshot();
    });
  });
});
