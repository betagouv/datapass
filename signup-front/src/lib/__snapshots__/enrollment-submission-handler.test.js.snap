// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When submitting the enrollment form with the destroy action calls the delete endpoint 1`] = `
Object {
  "errorMessages": Array [],
  "redirectToHome": true,
  "successMessages": Array [],
}
`;

exports[`When submitting the enrollment form with the notify action calls for the enrollment state update 1`] = `
Object {
  "errorMessages": Array [],
  "redirectToHome": true,
  "successMessages": Array [],
}
`;

exports[`When submitting the enrollment form with the update action calls the update endpoint 1`] = `
Object {
  "errorMessages": Array [],
  "redirectToHome": false,
  "successMessages": Array [
    "Votre demande a été sauvegardée.",
  ],
}
`;

exports[`When submitting the enrollment form with the update action displays an error if update fails 1`] = `
Object {
  "errorMessages": Array [
    "Une erreur inconnue est survenue. Merci de réessayer ultérieurement. Vous pouvez également nous signaler cette erreur par mail à contact@api.gouv.fr.",
  ],
  "redirectToHome": false,
  "successMessages": Array [],
}
`;
